app/src/main/java/com/babylon/wallet/android/data/dapp/IncomingRequestRepository.kt

```diff
@@ -68,25 +72,71 @@ class IncomingRequestRepositoryImpl @Inject constructor() : IncomingRequestRepos
         }
     }
 
+    /**
+     * There are two path of execution when using this method:
+     * - there is high priority screen in the queue, so the incoming request is added below it,
+     * taking priority over requests currently in the queue
+     * - there is no high priority screen: request is added at the top of queue and if there other request currently handled,
+     * we send a defer event for it so that UI can react and defer handling, without removing it from the queue.
+     * Deferred request will be handled again when top priority one handling completes
+     */
+    override suspend fun addMobileConnectRequest(incomingRequest: IncomingRequest) {
+        mutex.withLock {
+            requestQueue.addFirst(QueueItem.RequestItem(incomingRequest))
+            val currentRequest = _currentRequestToHandle.value
+            val handlingPaused = requestQueue.contains(QueueItem.HighPriorityScreen)
+            when {
+                currentRequest != null -> {
+                    Timber.d("🗂 Deferring request with id ${currentRequest.interactionId}")
+                    appEventBus.sendEvent(AppEvent.DeferRequestHandling(currentRequest.interactionId))
+                }
+
+                else -> {
+                    if (!handlingPaused) {
+                        handleNextRequest()
+                    }
+                }
+            }
+        }
+    }
+
     override suspend fun requestHandled(requestId: String) {
         mutex.withLock {
-            requestQueue.removeIf { it is QueueItem.RequestItem && it.incomingRequest.interactionId.toString() == requestId }
+            requestQueue.removeIf { it is QueueItem.RequestItem && it.incomingRequest.interactionId == requestId }
+            clearCurrent(requestId)
+            handleNextRequest()
+            Timber.d("🗂 request $requestId handled so size of list is now: ${getAmountOfRequests()}")
+        }
+    }
+
+    override suspend fun requestDeferred(requestId: String) {
```

Comment:
It is not actually deferred here? This is like a yield action right?

Reply:
It is deferred, but this method works in tandem with `addMobileConnectRequest`. When you use the latter, new request is put in front, and calling `requestDeferred` is simply taking next request to handle without removing current from the queue. I couldn't do both those actions in one method as there are user interactions in between.